### PR TITLE
Add Bitbucket provider.

### LINKED ISSUE #40: BitBucket support
I work for an org that uses BitBucket Cloud for all of our repos. It'd be nice if oauth2_proxy could be used w/ BitBucket.

## Expected Behavior

It'd be nice if oauth2_proxy has a BitBucket provider.

## Current Behavior

oauth2_proxy can't be used with BitBucket right now. BitBucket supports Oauth2, but not OpenID connect, so a new provider would be needed.

## Possible Solution

Add bitbucket as a provider? I'd be willing to spend some time writing a PR for this, but it may take me awhile and I lack confidence in my skills.

## Steps to Reproduce (for bugs)

N/A

## Context

The org I work for uses BitBucket cloud for all of our repos. I'd like to deploy various internal tools for developers to access and authorize them using their BitBucket credentials.

### PR DESCRIPTION
**Pull request by aledeganopix4d**
##  Description

Adds Bitbucket to the providers, includes tests.

## Motivation and Context

Fix #40 

## How Has This Been Tested?

This PR is based on a fork of the original Bitly project. We've been using that fork in PROD for several months by now.

## Checklist:

- [X] My change requires a change to the documentation or CHANGELOG.
- [X] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: providers/bitbucket.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+// SetRepository defines the repository the user must have access to
+func (p *BitbucketProvider) SetRepository(repository string) {
+	p.Repository = repository
+	if !strings.Contains(p.Scope, "repository") {
+		p.Scope += " repository"
+	}
+}
+
+func debug(data []byte, err error) {
```
Comment (JoelSpeed) [inline providers/bitbucket.go]: This is different to how we handle errors normally, what's the benefit of this?

File: providers/bitbucket.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	}
+	err = api.RequestJSON(req, &emails)
+	if err != nil {
+		log.Printf("failed making request %s", err)
+		debug(httputil.DumpRequestOut(req, true))
+		return "", err
+	}
+
+	if p.Team != "" {
+		log.Printf("Filtering against membership in team %s\n", p.Team)
```
Comment (JoelSpeed) [inline providers/bitbucket.go]: Does this not get very noisy when logging for each authentication attempt?

File: providers/bitbucket.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+			return "", err
+		}
+		err = api.RequestJSON(req, &teams)
+		if err != nil {
+			log.Printf("failed requesting teams membership %s", err)
+			debug(httputil.DumpRequestOut(req, true))
+			return "", err
+		}
+		var found = false
+		log.Printf("%+v\n", teams)
```
Comment (JoelSpeed) [inline providers/bitbucket.go]: This looks like debug logging?

File: providers/bitbucket.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+			}
+		}
+		if found != true {
+			log.Printf("team membership test failed, access denied")
+			return "", nil
+		}
+	}
+
+	if p.Repository != "" {
+		log.Printf("Filtering against access to repository %s\n", p.Repository)
```
Comment (JoelSpeed) [inline providers/bitbucket.go]: Same as above, perhaps a bit noisy?

File: providers/bitbucket.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+			return "", err
+		}
+		err = api.RequestJSON(req, &repositories)
+		if err != nil {
+			log.Printf("failed checking repository access %s", err)
+			debug(httputil.DumpRequestOut(req, true))
+			return "", err
+		}
+		var found = false
+		log.Printf("%+v\n", repositories)
```
Comment (JoelSpeed) [inline providers/bitbucket.go]: Debug logging?

**Comment by aledeganopix4d** [comment]
Thank you @JoelSpeed and @steakunderscore for your review!
I think I've addressed all your comments.

File: options.go [hunk 6]
```diff
@@ -41,6 +41,8 @@ type Options struct {
 
 	AuthenticatedEmailsFile  string   `flag:"authenticated-emails-file" cfg:"authenticated_emails_file" env:"OAUTH2_PROXY_AUTHENTICATED_EMAILS_FILE"`
 	AzureTenant              string   `flag:"azure-tenant" cfg:"azure_tenant" env:"OAUTH2_PROXY_AZURE_TENANT"`
+	BitbucketTeam            string   `flag:"bitbucket-team" cfg:"bitbucket_team" env:"OAUTH2_PROXY_BITBUCKET_TEAM"`
```
Comment (steakunderscore) [inline options.go]: Where does `BitbucketTeam` get read? Seems like there's a link missing between this line and `SetTeam()` being called. Did you test that setting the team works?

File: options.go [hunk 7]
```diff
@@ -41,6 +41,8 @@ type Options struct {
...
+	BitbucketRepository      string   `flag:"bitbucket-repository" cfg:"bitbucket_repository" env:"OAUTH2_PROXY_BITBUCKET_REPOSITORY"`
```
Comment (steakunderscore) [inline options.go]: Same as `BitbucketTeam`

File: providers/bitbucket.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+		}
+	}
+	if p.Scope == "" {
+		p.Scope = "email"
+	}
+	return &BitbucketProvider{ProviderData: p}
+}
+
+// SetTeam defines the Bitbucket team the user must be part of
+func (p *BitbucketProvider) SetTeam(team string) {
```
Comment (steakunderscore) [inline providers/bitbucket.go]: Same issue as above. Where does `SetTeam()` get called from?

File: options.go (hunk 6 above)
Comment (aledeganopix4d) [reply options.go]: I'm sorry. I've ported this PR from some code we had on a fork of the original `oauth2_proxy` from bitly.
Some bits got lost in translation. I've now fixed this and tested on a prod machine wrt teams and repos.

**Comment by aledeganopix4d** [comment]
I've retested the PR in prod, note that #215 was necessary to properly test, but I've split the fix in its own PR.
Seems to be working as expected.

**Comment by YusukeIwaki** [comment]
It would be better to define `GetUserName`, like GitHub provider.

> GetUserName has been added to the Provider interface to get User value when not equal to Email's local-part.

(from https://github.com/bitly/oauth2_proxy/pull/466)

The same may be said of BitBucket account.

**Comment by steakunderscore** [comment]
> It would be better to define `GetUserName`, like GitHub provider.

Agreed it would be useful to define `GetUserName`, but I would fine with that being added in separate PR as it's not a requirement.

**Comment by davidholsgrove** [comment]
Tested against our Bitbucket org and this looks great @aledeganopix4d 
Looking forward to this being merged into master / a release 👍

**Pull request merged**